File: scripts/ibkr-option-quote.py
#!/usr/bin/env python3
"""
Fetch live IBKR option quotes for a multi-leg structure via ib_insync.

Connects to IB Gateway (port 4001) or TWS (port 7496).

Usage:
  python3 scripts/ibkr-option-quote.py IBIT "BUY 49C 20260821, BUY 55C 20260821, SELL 60C 20260821 x2, SELL 36P 20260821"

Leg format: ACTION STRIKE[C|P] EXPIRY [xQTY]
"""

import sys
import re

# Use Radon's venv which has ib_insync
sys.path.insert(0, "/Users/home-hub/projects/radon/.venv/lib/python3.14/site-packages")

from lib.ibkr_option_quote_boundary import (  # noqa: E402
    INTERACTIVE_QUOTE_CLIENT_ID,
    positive_number,
    qualify_requested_option,
)


def parse_legs(text):
    """Parse leg string into structured list."""
    legs = []
    for part in text.split(","):
        part = part.strip()
        m = re.match(r"(BUY|SELL)\s+(\d+(?:\.\d+)?)(C|P)\s+(\d{8})(?:\s+x(\d+))?", part, re.I)
        if not m:
            print(f"  Cannot parse: '{part}'")
            continue
        legs.append({
            "action": m.group(1).upper(),
            "strike": float(m.group(2)),
            "right": m.group(3).upper(),
            "expiry": m.group(4),
            "qty": int(m.group(5)) if m.group(5) else 1,
        })
    return legs


def connect_quote_gateway(ib):
    """Connect for quotes without acquiring gateway lifecycle authority."""
    for port in (4001, 7496):
        try:
            ib.connect(
                "127.0.0.1",
                port,
                clientId=INTERACTIVE_QUOTE_CLIENT_ID,
                timeout=20,
            )
            return True, None
        except Exception:
            continue
    return False, "gateway-unavailable"


def main():
    from ib_insync import IB, Option  # noqa: E402

    if len(sys.argv) < 3:
        print("Usage: python3 scripts/ibkr-option-quote.py TICKER \"BUY 49C 20260821, SELL 60C 20260821 x2, ...\"")
        return 1

    ticker = sys.argv[1].upper()
    legs = parse_legs(sys.argv[2])

    if not legs:
        print("No valid legs parsed")
        return 1

    print(f"\nIBKR Live Quote: {ticker}")
    print(f"Legs: {len(legs)}")
    for l in legs:
        print(f"  {l['action']} {l['qty']}x {l['strike']}{l['right']} {l['expiry']}")

    # Connect to IB Gateway
    ib = IB()
    print(f"\nConnecting to IB Gateway...")

    connected, unavailable_reason = connect_quote_gateway(ib)
    if not connected:
        print(f"  UNAVAILABLE: {unavailable_reason}")
        print("  Gateway lifecycle recovery belongs to the separate /gateway workflow")
        return 2

    print(f"  Connected: {ib.managedAccounts()}")

    # Live data by default (nick gateway profile carries the streaming bundle);
    # --delayed forces type 3 for sessions on an unentitled login
    if "--delayed" in sys.argv:
        ib.reqMarketDataType(3)  # 3 = delayed, 4 = delayed-frozen
        print("  Market data type: DELAYED (forced)")
    else:
        ib.reqMarketDataType(1)  # 1 = live; unentitled contracts just won't populate
        print("  Market data type: LIVE (use --delayed if quotes come back empty)")

    # Build option contracts
    contracts = []
    for l in legs:
        request = {
            "ticker": ticker,
            "expiry": l["expiry"],
            "strike": l["strike"],
            "right": l["right"],
        }
        contract, reason = qualify_requested_option(
            ib,
            request,
            Option,
        )
        contracts.append((l, contract))
        if contract is not None:
            print(f"  {l['action']} {l['qty']}x {l['strike']}{l['right']} → conId={contract.conId}")
        else:
            print(f"  {l['action']} {l['qty']}x {l['strike']}{l['right']} → UNAVAILABLE: {reason}")

    # Request market data
    print(f"\nFetching live quotes...")
    tickers_map = {}
    for leg, contract in contracts:
        if contract is not None and contract.conId:
            try:
                t = ib.reqMktData(contract, "", False, False)
            except Exception:
                continue
            tickers_map[contract.conId] = (leg, contract, t)

    # Wait for data to populate
    ib.sleep(3)

    # Display results
    print(f"\n{'='*95}")
    print(f"IBKR LIVE QUOTES: {ticker}")
    print(f"{'='*95}")
    print(f"{'Leg':<6} {'Strike':<8} {'Expiry':<10} {'Qty':>4} {'Bid':>8} {'Ask':>8} {'Mid':>8} {'Last':>8} {'IV':>8}")
    print("-" * 95)

    net_bid = 0  # cost to BUY the combo
    net_ask = 0
    net_mid = 0
    all_have_quotes = len(tickers_map) == len(contracts)

    for leg, contract, ticker_data in tickers_map.values():
        # Try live first, fall back to delayed
        bid = positive_number(ticker_data.bid)
        ask = positive_number(ticker_data.ask)
        last = positive_number(ticker_data.last)

        # Delayed data fields
        if bid is None and hasattr(ticker_data, 'delayedBid'):
            bid = positive_number(ticker_data.delayedBid)
        if ask is None and hasattr(ticker_data, 'delayedAsk'):
            ask = positive_number(ticker_data.delayedAsk)
        if last is None and hasattr(ticker_data, 'delayedLast'):
            last = positive_number(ticker_data.delayedLast)

        mid = (bid + ask) / 2 if bid is not None and ask is not None else None

        # IV from model greeks or delayed greeks
        iv = None
        if hasattr(ticker_data, 'modelGreeks') and ticker_data.modelGreeks:
            iv = ticker_data.modelGreeks.impliedVol

        # Combo pricing
        if bid is not None and ask is not None:
            if leg["action"] == "BUY":
                net_ask += ask * leg["qty"]
                net_bid += bid * leg["qty"]
            else:
                net_ask -= bid * leg["qty"]
                net_bid -= ask * leg["qty"]
            sign = 1 if leg["action"] == "BUY" else -1
            net_mid += sign * mid * leg["qty"]
        else:
            all_have_quotes = False

        qty_str = f"x{leg['qty']}" if leg['qty'] > 1 else ""
        bid_s = f"${bid:.2f}" if bid else "—"
        ask_s = f"${ask:.2f}" if ask else "—"
        mid_s = f"${mid:.2f}" if mid else "—"
        last_s = f"${last:.2f}" if last else "—"
        iv_s = f"{iv:.1%}" if iv else "—"

        print(f"{leg['action']:<6} {leg['strike']}{leg['right']:<6} {leg['expiry']:<10} {qty_str:>4} {bid_s:>8} {ask_s:>8} {mid_s:>8} {last_s:>8} {iv_s:>8}")

    print("-" * 95)

    if all_have_quotes:
        print(f"{'COMBO':<6} {'':14} {'':10} {'':4} ${net_bid:>7.2f} ${net_ask:>7.2f} ${net_mid:>7.2f}")
        print()
        combo_type = "DEBIT" if net_mid > 0 else "CREDIT"
        print(f"Combo Net: ${abs(net_mid):.2f} {combo_type} at mid")
        print(f"  Bid: ${net_bid:.2f} | Ask: ${net_ask:.2f} | Spread: ${net_ask - net_bid:.2f}")
        per_ct = abs(net_mid) * 100
        print(f"  Per contract: ${per_ct:.0f} | 30 contracts: ${per_ct * 30:,.0f}")
    else:
        print("\n  UNAVAILABLE: market-data-unavailable-or-contract-unqualified")

    # Cancel market data
    for _, _, t in tickers_map.values():
        ib.cancelMktData(t.contract)

    ib.disconnect()
    return 0 if all_have_quotes else 2


if __name__ == "__main__":
    sys.exit(main())

File: scripts/ibkr-option-quote.ts
#!/usr/bin/env tsx
/**
 * ⚠ DEPRECATED (2026-07-10) — do not use for live verification.
 *
 * This is the OLD Client Portal Gateway (port 5001) quote path. The CP gateway
 * is a legacy livePrices fallback and is usually NOT running — its absence is
 * normal. The live-verification path is the always-on IBC/TWS gateway on port
 * 4001 (launchd local.ibc-gateway, docs/v2/21):
 *   - multi-leg human-readable:  scripts/ibkr-option-quote.py  (the /ibkr-quote skill)
 *   - batch machine-readable:    scripts/ibkr-quote-contracts.py (stdin JSON)
 * Kept only for the rare case the CP gateway is deliberately running.
 *
 * Fetch live IBKR option quotes for a multi-leg structure.
 *
 * Uses the IBKR Client Portal Gateway API to get real bid/ask/last
 * for each leg, then computes the combo net price.
 *
 * Prerequisites:
 *   - IBKR Client Portal Gateway running (default https://localhost:5001)
 *   - Gateway authenticated (log in via browser)
 *
 * Usage:
 *   npx tsx scripts/ibkr-option-quote.ts --ticker IBIT --legs "BUY 49C Aug21, SELL 60C Aug21 x2, BUY 55C Aug21, SELL 36P Aug21"
 *   npx tsx scripts/ibkr-option-quote.ts --ticker IBIT --legs "BUY 49C 20260821, SELL 60C 20260821 x2, BUY 55C 20260821, SELL 36P 20260821"
 *
 * Leg format: ACTION STRIKE[C|P] EXPIRY [xQTY]
 *   ACTION: BUY or SELL
 *   STRIKE: number + C (call) or P (put)
 *   EXPIRY: YYYYMMDD or MonDD (e.g., Aug21 → 20260821)
 *   xQTY: optional multiplier (default 1)
 */

import { config } from 'dotenv';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
config({ path: join(__dirname, '..', '.env.local') });

// ===================== IBKR CLIENT =====================

const GATEWAY_URL = process.env.IBKR_GATEWAY_URL || 'https://localhost:5001';

async function ibkrGet<T>(endpoint: string): Promise<T> {
  const origReject = process.env.NODE_TLS_REJECT_UNAUTHORIZED;
  process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
  try {
    const resp = await fetch(`${GATEWAY_URL}${endpoint}`, {
      headers: { 'Content-Type': 'application/json' },
    });
    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`IBKR ${resp.status}: ${text.slice(0, 200)}`);
    }
    return await resp.json();
  } finally {
    if (origReject !== undefined) process.env.NODE_TLS_REJECT_UNAUTHORIZED = origReject;
    else delete process.env.NODE_TLS_REJECT_UNAUTHORIZED;
  }
}

async function ibkrPost<T>(endpoint: string, body: any): Promise<T> {
  const origReject = process.env.NODE_TLS_REJECT_UNAUTHORIZED;
  process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
  try {
    const resp = await fetch(`${GATEWAY_URL}${endpoint}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });
    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`IBKR ${resp.status}: ${text.slice(0, 200)}`);
    }
    return await resp.json();
  } finally {
    if (origReject !== undefined) process.env.NODE_TLS_REJECT_UNAUTHORIZED = origReject;
    else delete process.env.NODE_TLS_REJECT_UNAUTHORIZED;
  }
}

// ===================== LEG PARSING =====================

interface ParsedLeg {
  action: 'BUY' | 'SELL';
  strike: number;
  right: 'C' | 'P';
  expiry: string; // YYYYMMDD
  qty: number;
}

function parseExpiry(raw: string): string {
  // If already YYYYMMDD
  if (/^\d{8}$/.test(raw)) return raw;

  // MonDD or MonYY format: Aug21 → 20260821
  const months: Record<string, string> = {
    Jan: '01', Feb: '02', Mar: '03', Apr: '04', May: '05', Jun: '06',
    Jul: '07', Aug: '08', Sep: '09', Oct: '10', Nov: '11', Dec: '12',
  };
  const match = raw.match(/^([A-Za-z]{3})(\d{1,2})(?:'?(\d{2}))?$/);
  if (match) {
    const mon = months[match[1]!.charAt(0).toUpperCase() + match[1]!.slice(1).toLowerCase()];
    const day = match[2]!.padStart(2, '0');
    const year = match[3] ? `20${match[3]}` : '2026'; // default to current year context
    if (mon) return `${year}${mon}${day}`;
  }

  throw new Error(`Cannot parse expiry: ${raw}`);
}

function parseLegs(input: string): ParsedLeg[] {
  return input.split(',').map(part => {
    part = part.trim();
    // Match: BUY 49C Aug21 x2
    const m = part.match(/^(BUY|SELL)\s+(\d+(?:\.\d+)?)(C|P)\s+(\S+)(?:\s+x(\d+))?$/i);
    if (!m) throw new Error(`Cannot parse leg: "${part}". Expected: BUY 49C Aug21 [x2]`);

    return {
      action: m[1]!.toUpperCase() as 'BUY' | 'SELL',
      strike: parseFloat(m[2]!),
      right: m[3]!.toUpperCase() as 'C' | 'P',
      expiry: parseExpiry(m[4]!),
      qty: m[5] ? parseInt(m[5]) : 1,
    };
  });
}

// ===================== IBKR SECDEF / CHAIN =====================

interface SecDefResult {
  conid: number;
  symbol: string;
  secType: string;
  exchange: string;
  listingExchange?: string;
  sections?: Array<{ secType: string; exchange?: string; conid?: string }>;
}

interface StrikesResult {
  call: number[];
  put: number[];
}

async function getUnderlyingConid(ticker: string): Promise<number> {
  const results = await ibkrGet<SecDefResult[]>(
    `/v1/api/iserver/secdef/search?symbol=${encodeURIComponent(ticker)}&sectype=STK`
  );
  if (!results || results.length === 0) throw new Error(`No contract found for ${ticker}`);

  // Prefer primary listing
  const stock = results.find(r => r.sections?.some(s => s.secType === 'STK'));
  const conid = stock?.conid || results[0]!.conid;
  return conid;
}

async function getOptionConid(
  underlyingConid: number,
  expiry: string,
  strike: number,
  right: 'C' | 'P'
): Promise<number | null> {
  // Use secdef/info to get specific option contract
  const params = new URLSearchParams({
    conid: String(underlyingConid),
    sectype: 'OPT',
    month: expiry.slice(0, 6), // YYYYMM
    right: right,
    strike: String(strike),
    exchange: '',
  });

  try {
    const results = await ibkrGet<any[]>(
      `/v1/api/iserver/secdef/info?${params}`
    );
    if (results && results.length > 0) {
      // Find exact match
      const match = results.find((r: any) =>
        r.maturityDate === expiry &&
        Math.abs(parseFloat(r.strike) - strike) < 0.01 &&
        r.right === right
      );
      return match?.conid || results[0]?.conid || null;
    }
  } catch (err) {
    console.error(`  Failed to find option ${strike}${right} ${expiry}:`, err instanceof Error ? err.message : String(err));
  }
  return null;
}

// ===================== MARKET DATA =====================

interface MarketDataSnapshot {
  conid: number;
  '31'?: string; // last
  '84'?: string; // bid
  '86'?: string; // ask
  '7283'?: string; // IV
  '7633'?: string; // IV at strike
  [key: string]: any;
}

async function getMarketDataSnapshot(conids: number[]): Promise<MarketDataSnapshot[]> {
  // Request fields: 31=last, 84=bid, 86=ask, 7283=IV
  const fields = '31,84,86,7283';
  const results = await ibkrGet<MarketDataSnapshot[]>(
    `/v1/api/iserver/marketdata/snapshot?conids=${conids.join(',')}&fields=${fields}`
  );

  // IBKR often needs a second request to populate data
  await new Promise(r => setTimeout(r, 1500));

  const retry = await ibkrGet<MarketDataSnapshot[]>(
    `/v1/api/iserver/marketdata/snapshot?conids=${conids.join(',')}&fields=${fields}`
  );

  return retry || results;
}

function parsePrice(val: string | undefined): number | null {
  if (!val) return null;
  // IBKR sometimes prefixes with C/H for close/halted
  const cleaned = val.replace(/^[A-Z]/, '');
  const num = parseFloat(cleaned);
  return isNaN(num) ? null : num;
}

// ===================== FORMAT HELPERS =====================

const pad = (s: string, n: number) => s.padEnd(n);
const rpad = (s: string, n: number) => s.padStart(n);

// ===================== MAIN =====================

async function main() {
  const args = process.argv.slice(2);
  const tickerIdx = args.indexOf('--ticker');
  const legsIdx = args.indexOf('--legs');

  if (tickerIdx < 0 || legsIdx < 0) {
    console.error('Usage: npx tsx scripts/ibkr-option-quote.ts --ticker IBIT --legs "BUY 49C Aug21, SELL 60C Aug21 x2, BUY 55C Aug21, SELL 36P Aug21"');
    process.exit(1);
  }

  const ticker = args[tickerIdx + 1]!.toUpperCase();
  const legsStr = args[legsIdx + 1]!;
  const legs = parseLegs(legsStr);

  console.log(`\nIBKR Live Quote: ${ticker}`);
  console.log(`Legs: ${legs.length}`);
  for (const l of legs) {
    console.log(`  ${l.action} ${l.qty}x ${l.strike}${l.right} ${l.expiry}`);
  }

  // Step 1: Verify gateway
  console.log('\nConnecting to IBKR Gateway...');
  try {
    await ibkrGet('/v1/api/tickle');
    console.log('  Gateway: connected');
  } catch (err) {
    console.error('  Gateway not available:', err instanceof Error ? err.message : String(err));
    console.error('  Make sure IBKR Client Portal Gateway is running and authenticated');
    process.exit(1);
  }

  // Step 2: Get underlying conid
  console.log(`\nResolving ${ticker} contract...`);
  const underlyingConid = await getUnderlyingConid(ticker);
  console.log(`  Underlying conid: ${underlyingConid}`);

  // Step 3: Resolve each leg's option conid
  console.log('\nResolving option contracts...');
  const legConids: Array<{ leg: ParsedLeg; conid: number | null }> = [];

  for (const leg of legs) {
    const conid = await getOptionConid(underlyingConid, leg.expiry, leg.strike, leg.right);
    legConids.push({ leg, conid });
    console.log(`  ${leg.action} ${leg.qty}x ${leg.strike}${leg.right} ${leg.expiry}: conid=${conid || 'NOT FOUND'}`);
  }

  const validConids = legConids.filter(lc => lc.conid !== null);
  if (validConids.length === 0) {
    console.error('\nNo option contracts could be resolved. Check expiry format (YYYYMMDD).');
    process.exit(1);
  }

  // Step 4: Fetch market data
  console.log('\nFetching live quotes...');
  const conidList = [...new Set(validConids.map(lc => lc.conid!))];
  const snapshots = await getMarketDataSnapshot(conidList);
  const snapMap = new Map(snapshots.map(s => [s.conid, s]));

  // Step 5: Display results
  console.log('\n' + '='.repeat(90));
  console.log(`IBKR LIVE QUOTES: ${ticker}`);
  console.log('='.repeat(90));
  console.log('Leg    Strike   Type  Expiry          Bid      Ask      Mid     Last       IV');
  console.log('-'.repeat(90));

  let netBid = 0; // to BUY the combo
  let netAsk = 0;
  let netMid = 0;

  for (const { leg, conid } of legConids) {
    if (!conid) {
      console.log(`${pad(leg.action, 6)} ${leg.strike}${leg.right}     —     ${pad(leg.expiry, 10)} ${rpad('N/A', 8)} ${rpad('N/A', 8)} ${rpad('N/A', 8)} ${rpad('N/A', 8)}`);
      continue;
    }

    const snap = snapMap.get(conid);
    const bid = parsePrice(snap?.['84']);
    const ask = parsePrice(snap?.['86']);
    const last = parsePrice(snap?.['31']);
    const iv = parsePrice(snap?.['7283']);
    const mid = bid !== null && ask !== null ? (bid + ask) / 2 : null;

    const sign = leg.action === 'BUY' ? 1 : -1;
    const qty = leg.qty;

    // Combo pricing: to BUY combo, pay ask on buys, receive bid on sells
    if (bid !== null && ask !== null) {
      if (leg.action === 'BUY') {
        netAsk += ask * qty;  // pay ask
        netBid += bid * qty;  // could sell at bid
      } else {
        netAsk -= bid * qty;  // receive bid when selling
        netBid -= ask * qty;  // would pay ask to buy back
      }
      netMid += sign * mid! * qty;
    }

    const bidStr = bid !== null ? `$${bid.toFixed(2)}` : '—';
    const askStr = ask !== null ? `$${ask.toFixed(2)}` : '—';
    const midStr = mid !== null ? `$${mid.toFixed(2)}` : '—';
    const lastStr = last !== null ? `$${last.toFixed(2)}` : '—';
    const ivStr = iv !== null ? `${(iv * 100).toFixed(1)}%` : '—';
    const qtyStr = qty > 1 ? `x${qty}` : '  ';

    console.log(`${pad(leg.action, 6)} ${leg.strike}${leg.right}${qtyStr}         ${pad(leg.expiry, 10)} ${rpad(bidStr, 8)} ${rpad(askStr, 8)} ${rpad(midStr, 8)} ${rpad(lastStr, 8)} ${rpad(ivStr, 8)}`);
  }

  console.log('-'.repeat(90));
  console.log(`COMBO                              ${rpad('$' + netBid.toFixed(2), 8)} ${rpad('$' + netAsk.toFixed(2), 8)} ${rpad('$' + netMid.toFixed(2), 8)}`);

  const comboType = netMid > 0 ? 'DEBIT' : 'CREDIT';
  console.log(`\nCombo Net: $${Math.abs(netMid).toFixed(2)} ${comboType} at mid`);
  console.log(`  Bid: $${netBid.toFixed(2)} | Ask: $${netAsk.toFixed(2)} | Spread: $${(netAsk - netBid).toFixed(2)}`);

  if (netMid !== 0) {
    const perContract = Math.abs(netMid) * 100;
    console.log(`  Per contract: $${perContract.toFixed(0)} | 30 contracts: $${(perContract * 30).toFixed(0)}`);
  }

  // Unsubscribe market data
  for (const conid of conidList) {
    try {
      await ibkrGet(`/v1/api/iserver/marketdata/${conid}/unsubscribe`);
    } catch {}
  }
}

main().catch(err => {
  console.error('Fatal:', err instanceof Error ? err.message : String(err));
  process.exit(1);
});
